Do not retrieve VM's stats on normal VM listing

JoaoJandre opened the pull request:
### Description

For both the `listVirtualMachines` and `listVirtualMachineMetrics` APIs, by default, ACS queries all VM details (the `details` parameter is default `all`) and consequently queries the `cloud.vm_stats` table for each VM, creating a summary of statistics. Therefore, every process carried out in the UI that uses the `listVirtualMachines` API also loads the VM statistics, which can cause slowdowns depending on the number of records.

A new configuration `return.vm.stats.on.vm.list` (with default true), was added, when it is false, the `listVirtualMachines` API will not list the VM's metrics by default.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Called listVirtualMachines API, without any parameters:
- before this PR, the all the VM's details were returned;
- with this PR and `return.vm.stats.on.vm.list` as true (default), all the details were returned as well;
- with this PR and `return.vm.stats.on.vm.list` as false, all the details except the metrics were returned.

weizhouapache commented:
@JoaoJandre 
I suggest to check/fix the API calls from UI to list vms or vm metrics.

It is not good to change the default behavior if it is not a bug, IMHO.

sureshanaparti commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/ListVMsCmd.java:
```diff
@@ -98,7 +99,8 @@ public class ListVMsCmd extends BaseListRetrieveOnlyResourceCountCmd implements
                description = "comma separated list of vm details requested, "
                    + "value can be a list of [all, group, nics, stats, secgrp, tmpl, servoff, diskoff, backoff, iso, volume, min, affgrp]."
-                   + " If no parameter is passed in, the details will be defaulted to all")
+                   + " If no parameter is passed in, the details will be defaulted to [group, nics, secgrp, tmpl, servoff, diskoff, iso, volume, min, affgrp] for single VM "
```
can use the relevant details in ui call to control the response, instead of updating the default behavior ?

JoaoJandre commented:
@blueorangutan package

blueorangutan commented:
@JoaoJandre a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9549

JoaoJandre commented:
@rohityadavcloud , @weizhouapache  I added a configuration to control the behavior of the `listVirtualMachines` API and updated the description. Now the default behavior is maintained. I hope we can advance the discussion on #8970 so we can create a proper mechanism to change the default behavior in the future.

weizhouapache commented:
@blueorangutan test rocky8 kvm-rocky8

weizhouapache commented:
> @rohityadavcloud , @weizhouapache I added a configuration to control the behavior of the `listVirtualMachines` API and updated the description. Now the default behavior is maintained. I hope we can advance the discussion on #8970 so we can create a proper mechanism to change the default behavior in the future.

thanks @JoaoJandre for the update

blueorangutan commented:
<b>[SF] Trillian test result (tid-10170)</b>
Environment: kvm-rocky8 (x2), Advanced Networking with Mgmt server r8
Total time taken: 49325 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8782-t10170-kvm-rocky8.zip
Smoke tests completed. 129 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_events_resource | `Error` | 418.29 | test_events_resource.py
test_01_restore_vm | `Error` | 0.24 | test_restore_vm.py
test_02_restore_vm_allocated_root | `Error` | 0.18 | test_restore_vm.py
ContextSuite context=TestRestoreVM>:teardown | `Error` | 1.26 | test_restore_vm.py

rohityadavcloud commented:
I like the idea of splitting which API is called for metrics vs non-metrics list view (I might or others should steal the pattern for all metrics API usage across the UI). I'm not fully satisfied with the PR @JoaoJandre yet, and I would rather encourage you can pick some ideas from #8985 - that said, I'm also inclined to make progress in a cordial and mature manner. It would be easier to get this merge and optimise the general solution further as required (by me or others).

I'm working on a much wider PR that's in research and progress, that's around wider scalability issues of CloudStack (surprisingly I'm near the root cause, and it maybe possible to even get stats without much penalty). All that said - I wouldn't remember everything I write on Github on each and every PR, take my review with a pinch of salt and I may change my views on things as I'm dealing with a wider scalability problem. I'll leave some comments, but think let's go ahead.

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/query/QueryService.java:
```diff
@@ -125,6 +125,10 @@ public interface QueryService {
     static final ConfigKey<Boolean> SharePublicTemplatesWithOtherDomains = new ConfigKey<>("Advanced", Boolean.class, "share.public.templates.with.other.domains", "true",
             "If false, templates of this domain will not show up in the list templates of other domains.", true, ConfigKey.Scope.Domain);
 
+    ConfigKey<Boolean> returnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "true",
```
Minor nit - the variable could be renamed:
 ```suggestion
    ConfigKey<Boolean> ReturnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "true",
```

See if there's a pattern in gobal setting name we should follow?

rohityadavcloud replied:
I'm wondering if the default can be set to `false`, if we do this we'll need to add a note in release notes about page - this breaks the default behaviour but given the listVirtualMachines API is widely used by CAPC, terraform, Ansible and what have you - other integrations could be faster. See https://github.com/apache/cloudstack/issues/8975

JoaoJandre replied:
I would really like to change the default behavior of this and some other APIs. But we currently do not have a proper mechanism to do this. We are slowly starting to discuss such mechanism in https://github.com/apache/cloudstack/discussions/8970#discussion-6554582, and more people discussing and giving ideas would be great.

In any case, since we are already creating a configuration for this, we might as well leave the default behavior as is and let users set this configuration if they feel the need to.

rohityadavcloud replied:
Generally speaking, the current mechanism of deprecation and change in behaviour that we've developed and used in recent years is through release notes documentation.

rohityadavcloud commented on plugins/metrics/src/main/java/org/apache/cloudstack/api/ListVMsMetricsCmd.java:
```diff
@@ -42,20 +44,24 @@
+        return super.getDetails();
     }
 
+
```
nit - remove extra line

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/ListVMsCmd.java:
```diff
@@ -239,22 +240,32 @@ public Long getAutoScaleVmGroupId() {
+                return EnumSet.of(VMDetails.all);
+            }
+
+            Set<VMDetails> allDetails = new HashSet<>(Set.of(VMDetails.values()));
```
I think it could be simpler, if the decision whether to list stats or not can be done via `server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java` 
```
-        if (details.contains(VMDetails.all) || details.contains(VMDetails.stats)) {
+        if (details.contains(VMDetails.stats)) {
```

See - https://github.com/apache/cloudstack/pull/8985/files#diff-b180fb2c499ee386bb5b8eae08c2a004e2c33a9430d94b4253bdbd3871d36f8a

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/api/response/UserVmResponse.java:
```diff
@@ -269,6 +270,10 @@ public class UserVmResponse extends BaseResponseWithTagInformation implements Co
 
+    @SerializedName(ApiConstants.IP_ADDRESS)
+    @Param(description = "the VM's primary IP address")
+    private String ipAddress;
```
Is there a way to avoid this, the ip address table can cause excessive scans in the DB.

JoaoJandre replied:
I don't see a reason to only return the primary IP on the metrics response, I think it should be returned when listing the user VMs as well. Furthermore, the metrics response extends the UserVmResponse, so when listing the metrics the IP will still be returned.

rohityadavcloud commented on plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java:
```diff
@@ -617,7 +617,6 @@ public List<VmMetricsResponse> listVmMetrics(List<UserVmResponse> vmResponses) {
             }
 
             metricsResponse.setHasAnnotation(vmResponse.hasAnnotation());
-            metricsResponse.setIpAddress(vmResponse.getNics());
```
Let's keep the metrics response as-is ? See remarks for listVirtualMachines API

JoaoJandre replied:
The IP address will come from the UserVmResponse, so the metrics response will be left as-is

rohityadavcloud commented on ui/src/config/section/compute.js:
```diff
@@ -28,7 +28,8 @@ export default {
       docHelp: 'adminguide/virtual_machines.html',
-      permission: ['listVirtualMachinesMetrics'],
+      permission: ['listVirtualMachines', 'listVirtualMachinesMetrics'],
+      getApiToCall: () => store.getters.metrics ? 'listVirtualMachinesMetrics' : 'listVirtualMachines',
```
Love this, we can use this across all places where we use metrics.

rohityadavcloud reviewed (approved):
Left some remarks, otherwise LGTM

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/api/command/user/vm/ListVMsCmd.java:
```diff
@@ -98,7 +99,7 @@ public class ListVMsCmd extends BaseListRetrieveOnlyResourceCountCmd implements
                description = "comma separated list of vm details requested, "
                    + "value can be a list of [all, group, nics, stats, secgrp, tmpl, servoff, diskoff, backoff, iso, volume, min, affgrp]."
-                   + " If no parameter is passed in, the details will be defaulted to all")
+                   + " If no parameter is passed in, the details will be defaulted to all.")
```
one more - could we mention in the API doc string here, about the global setting that determines whether stats will be sent or not by the list API?

DaanHoogland commented:
@sureshanaparti ok by you like this?

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✖️ debian ✔️ suse15. SL-JID 9631

DaanHoogland commented:
ping @sureshanaparti

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9663

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/query/QueryService.java:
```diff
@@ -125,6 +125,10 @@ public interface QueryService {
     static final ConfigKey<Boolean> SharePublicTemplatesWithOtherDomains = new ConfigKey<>("Advanced", Boolean.class, "share.public.templates.with.other.domains", "true",
             "If false, templates of this domain will not show up in the list templates of other domains.", true, ConfigKey.Scope.Domain);
 
+    ConfigKey<Boolean> ReturnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "true",
```
```suggestion
    ConfigKey<Boolean> ReturnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "false",
```

After due considerations, I think this can be set to false what originally Joao wants to accomplish. As the listVirtualMachines API is used by at least three integrations I know of (CAPC, k8s-provider and terraform etc.) and several portals are built on top such APIs, it makes sense to trade off speed. BUT, if we do this we must document this change in the release notes and what users can do to get previous/old behaviour if they want it.

What do you think - @JoaoJandre @weizhouapache @DaanHoogland @sureshanaparti @shwstppr ?

weizhouapache replied:
obviously users/other projects will benifit from the change, meanwhile it will break the backward compatibility. 
if we agree more pros than cons, let's go ahead. 
as @rohityadavcloud said, we must document in the release note and let users know (although many users might not be impacted and not be aware of the change).

rohityadavcloud replied:
We add the config for users who may want the backward compatibility. But we use an appropriate default that benefits most users.

blueorangutan commented:
<b>[SF] Trillian test result (tid-10254)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43718 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8782-t10254-kvm-centos7.zip
Smoke tests completed. 130 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_events_resource | `Error` | 416.93 | test_events_resource.py

DaanHoogland reviewed (approved):
I am ok with this given a short notice in the release notes. (cc @JoaoJandre @sureshanaparti )

rohityadavcloud commented on api/src/main/java/org/apache/cloudstack/query/QueryService.java:
```diff
@@ -125,6 +125,10 @@ public interface QueryService {
     static final ConfigKey<Boolean> SharePublicTemplatesWithOtherDomains = new ConfigKey<>("Advanced", Boolean.class, "share.public.templates.with.other.domains", "true",
             "If false, templates of this domain will not show up in the list templates of other domains.", true, ConfigKey.Scope.Domain);
 
+    ConfigKey<Boolean> ReturnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "true",
```
```suggestion
    ConfigKey<Boolean> ReturnVmStatsOnVmList = new ConfigKey<>("Advanced", Boolean.class, "return.vm.stats.on.vm.list", "false",
```

rohityadavcloud replied:
Looks like @JoaoJandre has disabled to allow maintainers to make changes 🤦 
Let's merge this and address this in subsequent PRs.
